pkg/volume/csi/csi_mounter.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"fmt"
+	"path"
+
+	"github.com/golang/glog"
+	grpctx "golang.org/x/net/context"
+	api "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type csiMountMgr struct {
+	csiClient  csiClient
+	plugin     *csiPlugin
+	driverName string
+	volumeID   string
+	readOnly   bool
+	spec       *volume.Spec
+	pod        *api.Pod
+	podUID     types.UID
+	options    volume.VolumeOptions
+	volume.MetricsNil
+}
+
+// volume.Volume methods
+var _ volume.Volume = &csiMountMgr{}
+
+func (c *csiMountMgr) GetPath() string {
+	driverPath := fmt.Sprintf("%s/%s", c.driverName, c.volumeID)
+	return getTargetPath(c.podUID, driverPath, c.plugin.host)
+}
+
+func getTargetPath(uid types.UID, driverPath string, host volume.VolumeHost) string {
+	return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(csiPluginName), driverPath)
+}
+
+// volume.Mounter methods
+var _ volume.Mounter = &csiMountMgr{}
+
+func (c *csiMountMgr) CanMount() error {
+	return nil
+}
+
+func (c *csiMountMgr) SetUp(fsGroup *int64) error {
+	return c.SetUpAt(c.GetPath(), fsGroup)
+}
+
+func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
+	glog.V(4).Infof(log("Mounter.SetUpAt(%s)", dir))
+
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	csi := c.csiClient
+
+	if err := csi.AssertSupportedVersion(ctx, csiVersion); err != nil {
+		glog.Errorf(log("failed to assert version: %v", err))
+		return err
+	}
+
+	accessMode := api.ReadWriteOnce
+	if c.spec.PersistentVolume.Spec.AccessModes != nil {
+		accessMode = c.spec.PersistentVolume.Spec.AccessModes[0]
```

Comment:
AccessMode is tricky. It is list of all modes that the volume supports. E.g. it can be all three `[ReadWriteOnce, ReadWriteMany, ReadOnlyMany]` for a NFS share - it can be used either as mounted to single pod, mounted to several pods or mounted to several pods as read only.

This gives us very little clue *how* the user wants to use the volume - is it a volume for MySQL database? Then it must be mounted only to one pod as multiple pods with MySQL will corrupt data of each other. Or should it be a common file share, where multiple pods access the same data? Kubernetes does not know it at this time. And therefore you should choose CSI capability that applies to all the PV modes.

See what attacher does: https://github.com/kubernetes-csi/external-attacher/blob/master/pkg/connection/util.go#L49
I am not very happy about this code either. CSI and Kubernetes are quite different in this aspect.

Reply:
Why not use the access mode specified in the PVC?

Reply:
PVC has the same problems. It's array and the volume can be used in any mode that the array specifies, one at the time. User can have a PVC with `[ReadWriteOne, ReadOnlyMany]`, writing some data to it in a single pod and then have multiple pods processing the data in read-only mode.

Reply:
@jsafrane Yes this is a point of confusion.  Thanks for the insight.

Reply:
@jsafrane I thought `ReadWriteOnce` was automatically interpreted by k8s to  be write in one pod and read from many others ... Meaning you don't have to specify `ReadOnlyMany` in the array.  At least that's the way it's describe here in the k8s volume doc: 

>**Important!** A volume can only be mounted using one access mode at a time, even if it supports many. For example, a GCEPersistentDisk can be mounted as ReadWriteOnce by a single node or ReadOnlyMany by many nodes, but not at the same time.

It would seem only one accessMode can be specified at a time which would make this much easier.

Reply:
no, `ReadWriteOnce` is read + write by one pod. `ReadOnlyMany` is very exotic GCE feature, e.g. AWS does not have it.